aten/src/ATen/native/SegmentReduce.cpp

```diff
@@ -0,0 +1,81 @@
+#include <ATen/native/SegmentReduce.h>
+
+#include <ATen/ATen.h>
+
+#include <iostream>
+
+namespace at {
+namespace native {
+
+DEFINE_DISPATCH(segment_reduce_stub);
+
+enum ReductionType { MAX };
+const std::map<std::string, ReductionType> reduce2REDUCE = {
+    {"max", MAX},
+};
+
+Tensor _segment_reduce_cpu(
+    const Tensor& data,
+    std::string reduce,
+    const c10::optional<Tensor>& lengths,
+    const c10::optional<Tensor>& indices,
+    int64_t axis,
+    bool unsafe) {
+  TORCH_CHECK(axis == 0, "Currently only dim=0 is supported!");
+  TORCH_CHECK(data.dim() == 1);
+  TORCH_CHECK(!data.is_cuda());
+  TORCH_CHECK(data.numel() > 0);
+  TORCH_CHECK(
+      reduce2REDUCE.at(reduce) == MAX,
+      "Currently only 'max' reduction is supported!");
+
+  // length related checks
+  TORCH_CHECK(
+      lengths.has_value() && !indices.has_value(),
+      "Currently only lengths based reduction is supported!")
+  const auto& lengths_value = lengths.value();
+  TORCH_CHECK(lengths_value.dim() == 1);
+  TORCH_CHECK(data.get_device() == lengths_value.get_device());
+  TORCH_CHECK(data.dim() >= lengths_value.dim());
+
+  const auto lengths_contig = lengths_value.contiguous();
+  const auto data_contig = data.contiguous();
+
+  int64_t batch_size = lengths_contig.numel();
+  auto output = at::empty({batch_size}, data.options());
+
+  const auto* lengths_data = lengths_contig.data_ptr<int64_t>();
+  if (!unsafe) {
+    int64_t sum = 0;
+    for (int64_t i = 0; i < batch_size; ++i) {
+      sum += lengths_data[i];
+    }
+    TORCH_CHECK(sum == data.numel());
+  }
+
+  AT_DISPATCH_ALL_TYPES(
+      data_contig.scalar_type(), "_segment_reduce_cpu", ([&]() {
+        auto* output_data = output.data_ptr<scalar_t>();
+        const auto* values_data = data_contig.data_ptr<scalar_t>();
+        int64_t k = 0;
+        for (int64_t i = 0; i < batch_size; ++i) {
+          scalar_t reduction = std::numeric_limits<scalar_t>::lowest();
+          for (int64_t j = 0; j < lengths_data[i]; ++j) {
+            if (values_data[k] > reduction) {
```

Comment:
this does not properly handle nan values, you need something like
reduction = at::_isnan(data) ? data : std::max<scalar_t>(reduction,data); //std::max returns first arg if one of the args is nan

Reply:
Thanks for the pointer. Updated. I think you meant "at::_isnan(data) ? reduction ..." instead, right?

Reply:
It should be `data`, not `reduction`, to properly propagate `nan` (also, will be good to add a test).

Reply:
I actually updated the test in the PR with none value. I guess expected behaviour is not clear for me. If one of the values is None for a reduction, do we want to return None reduction or result based on remaining items within the segment?

I don't have preference one way or another but would like to keep similar convention as rest of PyTorch regarding none handling. If returning None is preferred, will update the test and PR.

Reply:
Default is propagate nan values, that is, if nan is encountered in the segment, it is returned. But the expression should be as in my first comment to return nan if either of the args is nan
```
reduction = at::_isnan(data) ? data : std::max<scalar_t>(reduction,data); //std::max returns first arg if one of the args is nan
```
If `data` is `nan` it is returned, otherwise if `reduction` is `nan` it will be returned as the first arg to `std::max`, otherwise max(reduction, data) is returned.